dxr/query.py

```diff
@@ -102,8 +102,27 @@ def results(self, offset=0, limit=100):
         # will OR the elements of the inner lists and then AND those OR balls
         # together.
         enabled_filters_by_name = filters_by_name(self.enabled_plugins)
-        filters = [[f(term, self.enabled_plugins) for f in enabled_filters_by_name[term['name']]]
-                   for term in self.terms]
+
+        def group_filters_by_term(predicate):
+            # Return a nested list of ES filters for each term, filtered on predicate(Filter).
+            return [[f(term, self.enabled_plugins) for f in enabled_filters_by_name[term['name']]
+                    if predicate(f)] for term in self.terms]
+
+        def group_filters_by_name(predicate):
+            # Return a nested list of ES filters for each unique filter name, filtered on
+            # predicate(Filter).
+            d = {}
+            for term in self.terms:
+                for f in enabled_filters_by_name[term['name']]:
+                    if predicate(f):
+                        d.setdefault(term['name'], []).append(f(term, self.enabled_plugins))
+            return d.values()
+
+        # Some filters, such as ExtFilter, do not make sense to be AND'ed together, so we move
+        # them all to their own lists at the end of the regular filters list, such that they
+        # will be joined by OR instead.
+        filters = (group_filters_by_term(lambda f: not f.union_only) +
```

Comment:
Probably not an issue in practical terms, but we'd likely see a little efficiency boost if we could avoid +ing lists together. We could make the two group_filters_by_* functions into generators and list(the first) and the_list.extend(the second).